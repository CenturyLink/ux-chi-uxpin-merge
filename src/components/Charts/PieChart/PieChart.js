/** @jsx jsx */
import { jsx } from '@emotion/core';
import PropTypes from 'prop-types';
import React from 'react';
import {
  Hint,
  RadialChart,
} from 'react-vis';
import ChartStyles from '../chartStyles/chart.styles';

export default class PieChart extends React.Component {
  constructor(props) {
    super(props);

    const getStartData = () => {
      if (Array.isArray(this.props.startData[0]) && this.props.startData.length === this.props.data.length) {
        return this.props.startData;
      } if (!Array.isArray(this.props.startData[0]) && this.props.startData.length === this.props.data.length) {
        return this.props.startData;
      }

      return this.props.data;
    };

    this.state = {
      data: this.props.startData ? getStartData() : this.props.data,
      hintValue: '',
      showHint: false,
    };
  }

  componentDidMount() {
    this.setState({ data: this.props.data });
  }

  getHint(value) {
    if (!this.state.showHint) {
      const hintContent = value.label ? { Data: value.label, Value: `${value.theta * 10}%` } : { Value: `${value.theta * 10}%` };
      this.setState({ hintValue: hintContent, showHint: true });
    } else {
      this.setState({ hintValue: '', showHint: false });
    }
  }

  render() {
    return (
      <RadialChart
        getAngle={(d) => d.theta}
        css={ChartStyles}
        radius={this.props.radius}
        innerRadius={this.props.innerRadius}
        padAngle={parseFloat(this.props.padAngle)}
        showLabels={this.props.showLabels}
        labelsRadiusMultiplier={parseFloat(this.props.labelsRadiusMultiplier)}
        labelsStyle={this.props.labelsStyle}
        data={this.state.data}
        width={this.props.width}
        height={this.props.height}
        colorRange={this.props.colorRange}
        opacity={parseFloat(this.props.opacity)}
        onValueClick={(value) => this.getHint(value)}
        onNearestXY={this.props.onNearestXY}
        onSeriesClick={this.props.onSeriesClick}
        onSeriesRightClick={this.props.onSeriesRightClick}
        onSeriesMouseOver={this.props.onSeriesMouseOver}
        onSeriesMouseOut={this.props.onSeriesMouseOver}
        animation={this.props.animation}>
        {this.props.hint && this.state.showHint ? (
          <Hint value={this.state.hintValue} />
        ) : (
          undefined
        )}
      </RadialChart>
    );
  }
}

/* eslint-disable sort-keys */
PieChart.propTypes = {
  /** Height of the Chart in px. Accepts only numbers. */
  height: PropTypes.number,
  /** Width of the Chart in px. Accepts only numbers. */
  width: PropTypes.number,
  /** Turns, on/off animation and allows for selection of different types of animations. */
  animation: PropTypes.oneOf([false, 'noWobble', 'gentle', 'wobbly', 'stiff']),
  /**
   * Color to be used on all chart lines, unless colorRange is provided
   * @uxpinignoreprop
   */
  // eslint-disable-next-line react/no-unused-prop-types
  color: PropTypes.string,
  /** Array with colors to be used across all chart lines. If array doesn't specify color for all the chart lines, color property is used. */
  colorRange: PropTypes.arrayOf(PropTypes.string),
  /** Data Array. Structure:  [{ "theta": 1, "label": "apples" }, {"theta": 4, "label": "oranges"}, {"theta": 6, "label": "cherries"}]  */
  data: PropTypes.array,
  /** Shows hint on click into every part of the pie chart */
  hint: PropTypes.bool,
  /** In combination with radius property, innerRadius enables ability to create a donut chart. The higher the value the bigger tha hole in the donut.  */
  innerRadius: PropTypes.number,
  /** Specifies the distance between edge of the chart and labels. 1 places labels on the edge of the chart. */
  labelsRadiusMultiplier: PropTypes.string,
  /** Styles object that let's you style labels on the chart */
  labelsStyle: PropTypes.object,
  /**
   * @uxpinignoreprop
   */
  onNearestXY: PropTypes.func,
  /**
   * @uxpinignoreprop
   */
  onSeriesClick: PropTypes.func,
  /**
   * @uxpinignoreprop
   */
  // eslint-disable-next-line react/no-unused-prop-types
  onSeriesMouseOut: PropTypes.func,
  /**
   * @uxpinignoreprop
   */
  onSeriesMouseOver: PropTypes.func,
  /**
   * @uxpinignoreprop
   */
  onSeriesRightClick: PropTypes.func,
  /** Color filling inside of every circle on the chart. */
  opacity: PropTypes.string,
  /** Adds padding between series in the chart. 0 means no padding. 0.005 provides 1px relief. */
  padAngle: PropTypes.string,
  /** Specifies the radius of the chart. Value that equals 50% of the height of the entire component equals 100% of the size. Required to create donut chart. */
  radius: PropTypes.number,
  /** Shows labels one the chart. Set up labels in the data object. Example: {"theta": 1, "label": "Apples"} */
  showLabels: PropTypes.bool,
  /** Starting point for data set. Used for triggering animation. Same data structure as data property. Place "0" in theta to animate. */
  startData: PropTypes.array,
};
/* eslint-enable sort-keys */

PieChart.defaultProps = {
  height: 300,
  hint: false,
  labelsRadiusMultiplier: '1.1',
  opacity: '1',
  width: 300,
};
