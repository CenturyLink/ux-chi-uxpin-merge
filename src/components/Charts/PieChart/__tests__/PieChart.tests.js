import React from 'react';
import * as enzyme from 'enzyme';
import toJson from 'enzyme-to-json';
import PieChart from '../PieChart';


test('renders PieChart correctly', () => {
  const chart = (
    <PieChart
      width={300}
      height={300}
      data={[
        {
          label: 'apples',
          theta: 1,
        },
        {
          label: 'oranges',
          theta: 4,
        },
        {
          label: 'cherries',
          theta: 6,
        },
      ]}
      startData={[
        {
          label: 'apples',
          theta: 0,
        },
        {
          label: 'oranges',
          theta: 0,
        },
        {
          label: 'cherries',
          theta: 0,
        },
      ]}
      padAngle="0.02"
      radius={140}
      innerRadius={120}
    />
  );

  expect(toJson(enzyme.mount(chart))).toMatchSnapshot('enzyme.mount');
});
