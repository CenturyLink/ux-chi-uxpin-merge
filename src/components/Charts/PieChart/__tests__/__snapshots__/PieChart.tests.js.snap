// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders PieChart correctly: enzyme.mount 1`] = `
.emotion-0 .react-vis-magic-css-import-rule {
  display: inherit;
}

.emotion-0 .rv-treemap {
  font-size: 12px;
  position: relative;
}

.emotion-0 .rv-treemap__leaf {
  overflow: hidden;
  position: absolute;
}

.emotion-0 .rv-treemap__leaf--circle {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.emotion-0 .rv-treemap__leaf__content {
  overflow: hidden;
  padding: 10px;
  text-overflow: ellipsis;
}

.emotion-0 .rv-xy-plot {
  color: #c3c3c3;
  position: relative;
}

.emotion-0 .rv-xy-plot canvas {
  pointer-events: none;
}

.emotion-0 .rv-xy-plot .rv-xy-canvas {
  pointer-events: none;
  position: absolute;
}

.emotion-0 .rv-xy-plot__inner {
  display: block;
}

.emotion-0 .rv-xy-plot__axis__line {
  fill: none;
  stroke-width: 2px;
  stroke: #e6e6e9;
}

.emotion-0 .rv-xy-plot__axis__tick__line {
  stroke: #e6e6e9;
}

.emotion-0 .rv-xy-plot__axis__tick__text {
  fill: #6b6b76;
  font-size: 11px;
}

.emotion-0 .rv-xy-plot__axis__title text {
  fill: #6b6b76;
  font-size: 11px;
}

.emotion-0 .rv-xy-plot__grid-lines__line {
  stroke: #e6e6e9;
}

.emotion-0 .rv-xy-plot__circular-grid-lines__line {
  fill-opacity: 0;
  stroke: #e6e6e9;
}

.emotion-0 .rv-xy-plot__series,
.emotion-0 .rv-xy-plot__series path {
  pointer-events: all;
}

.emotion-0 .rv-xy-plot__series--line {
  fill: none;
  stroke: #000;
  stroke-width: 2px;
}

.emotion-0 .rv-crosshair {
  position: absolute;
  font-size: 11px;
  pointer-events: none;
}

.emotion-0 .rv-crosshair__line {
  background: #47d3d9;
  width: 1px;
}

.emotion-0 .rv-crosshair__inner {
  position: absolute;
  text-align: left;
  top: 0;
}

.emotion-0 .rv-crosshair__inner__content {
  border-radius: 4px;
  background: #3a3a48;
  color: #fff;
  font-size: 12px;
  padding: 7px 10px;
  box-shadow: 0 2px 4px rgba(0,0,0,0.5);
}

.emotion-0 .rv-crosshair__inner--left {
  right: 4px;
}

.emotion-0 .rv-crosshair__inner--right {
  left: 4px;
}

.emotion-0 .rv-crosshair__title {
  font-weight: bold;
  white-space: nowrap;
}

.emotion-0 .rv-crosshair__item {
  white-space: nowrap;
}

.emotion-0 .rv-hint {
  position: absolute;
  pointer-events: none;
}

.emotion-0 .rv-hint__content {
  border-radius: 4px;
  padding: 7px 10px;
  font-size: 12px;
  background: #3a3a48;
  box-shadow: 0 2px 4px rgba(0,0,0,0.5);
  color: #fff;
  text-align: left;
  white-space: nowrap;
}

.emotion-0 .rv-discrete-color-legend {
  box-sizing: border-box;
  overflow-y: auto;
  font-size: 12px;
}

.emotion-0 .rv-discrete-color-legend.horizontal {
  white-space: nowrap;
}

.emotion-0 .rv-discrete-color-legend-item {
  color: #3a3a48;
  border-radius: 1px;
  padding: 9px 10px;
}

.emotion-0 .rv-discrete-color-legend-item.horizontal {
  display: inline-block;
}

.emotion-0 .rv-discrete-color-legend-item.horizontal .rv-discrete-color-legend-item__title {
  margin-left: 0;
  display: block;
}

.emotion-0 .rv-discrete-color-legend-item__color {
  display: inline-block;
  vertical-align: middle;
  overflow: visible;
}

.emotion-0 .rv-discrete-color-legend-item__color__path {
  stroke: #dcdcdc;
  stroke-width: 2px;
}

.emotion-0 .rv-discrete-color-legend-item__title {
  margin-left: 10px;
}

.emotion-0 .rv-discrete-color-legend-item.disabled {
  color: #b8b8b8;
}

.emotion-0 .rv-discrete-color-legend-item.clickable {
  cursor: pointer;
}

.emotion-0 .rv-discrete-color-legend-item.clickable:hover {
  background: #f9f9f9;
}

.emotion-0 .rv-search-wrapper {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-0 .rv-search-wrapper__form {
  -webkit-flex: 0;
  -ms-flex: 0;
  flex: 0;
}

.emotion-0 .rv-search-wrapper__form__input {
  width: 100%;
  color: #a6a6a5;
  border: 1px solid #e5e5e4;
  padding: 7px 10px;
  font-size: 12px;
  box-sizing: border-box;
  border-radius: 2px;
  margin: 0 0 9px;
  outline: 0;
}

.emotion-0 .rv-search-wrapper__contents {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  overflow: auto;
}

.emotion-0 .rv-continuous-color-legend {
  font-size: 12px;
}

.emotion-0 .rv-continuous-color-legend .rv-gradient {
  height: 4px;
  border-radius: 2px;
  margin-bottom: 5px;
}

.emotion-0 .rv-continuous-size-legend {
  font-size: 12px;
}

.emotion-0 .rv-continuous-size-legend .rv-bubbles {
  text-align: justify;
  overflow: hidden;
  margin-bottom: 5px;
  width: 100%;
}

.emotion-0 .rv-continuous-size-legend .rv-bubble {
  background: #d8d9dc;
  display: inline-block;
  vertical-align: bottom;
}

.emotion-0 .rv-continuous-size-legend .rv-spacer {
  display: inline-block;
  font-size: 0;
  line-height: 0;
  width: 100%;
}

.emotion-0 .rv-legend-titles {
  height: 16px;
  position: relative;
}

.emotion-0 .rv-legend-titles__left,
.emotion-0 .rv-legend-titles__right,
.emotion-0 .rv-legend-titles__center {
  position: absolute;
  white-space: nowrap;
  overflow: hidden;
}

.emotion-0 .rv-legend-titles__center {
  display: block;
  text-align: center;
  width: 100%;
}

.emotion-0 .rv-legend-titles__right {
  right: 0;
}

.emotion-0 .rv-radial-chart .rv-xy-plot__series--label {
  pointer-events: none;
}

<PieChart
  data={
    Array [
      Object {
        "label": "apples",
        "theta": 1,
      },
      Object {
        "label": "oranges",
        "theta": 4,
      },
      Object {
        "label": "cherries",
        "theta": 6,
      },
    ]
  }
  height={300}
  hint={false}
  innerRadius={120}
  labelsRadiusMultiplier="1.1"
  opacity="1"
  padAngle="0.02"
  radius={140}
  startData={
    Array [
      Object {
        "label": "apples",
        "theta": 0,
      },
      Object {
        "label": "oranges",
        "theta": 0,
      },
      Object {
        "label": "cherries",
        "theta": 0,
      },
    ]
  }
  width={300}
>
  <RadialChart
    className="emotion-0"
    colorRange={
      Array [
        "#12939A",
        "#79C7E3",
        "#1A3177",
        "#FF9833",
        "#EF5D28",
      ]
    }
    colorType="category"
    data={
      Array [
        Object {
          "label": "apples",
          "theta": 1,
        },
        Object {
          "label": "oranges",
          "theta": 4,
        },
        Object {
          "label": "cherries",
          "theta": 6,
        },
      ]
    }
    getAngle={[Function]}
    getAngle0={[Function]}
    getLabel={[Function]}
    getRadius={[Function]}
    getRadius0={[Function]}
    getSubLabel={[Function]}
    height={300}
    innerRadius={120}
    labelsRadiusMultiplier={1.1}
    onValueClick={[Function]}
    opacity={1}
    padAngle={0.02}
    radius={140}
    width={300}
  >
    <XYPlot
      className="emotion-0 rv-radial-chart"
      height={300}
      margin={
        Object {
          "bottom": 10,
          "left": 10,
          "right": 10,
          "top": 10,
        }
      }
      width={300}
      xDomain={
        Array [
          -1,
          1,
        ]
      }
      yDomain={
        Array [
          -1,
          1,
        ]
      }
    >
      <div
        className="rv-xy-plot emotion-0 rv-radial-chart"
        style={
          Object {
            "height": "300px",
            "width": "300px",
          }
        }
      >
        <svg
          className="rv-xy-plot__inner"
          height={300}
          onClick={[Function]}
          onDoubleClick={[Function]}
          onMouseDown={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onMouseMove={[Function]}
          onMouseUp={[Function]}
          onTouchCancel={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          onWheel={[Function]}
          width={300}
        >
          <ArcSeries
            _adjustBy={Array []}
            _adjustWhat={Array []}
            _allData={
              Array [
                Array [
                  Object {
                    "angle": 6.283185307179586,
                    "angle0": 2.8559933214452666,
                    "color": 0,
                    "label": "cherries",
                    "radius": 1,
                    "radius0": 0,
                    "theta": 6,
                    "x": 1,
                    "y": -2.4492935982947064e-16,
                  },
                  Object {
                    "angle": 2.8559933214452666,
                    "angle0": 0.5711986642890533,
                    "color": 1,
                    "label": "oranges",
                    "radius": 1,
                    "radius0": 0,
                    "theta": 4,
                    "x": -0.9594929736144974,
                    "y": 0.28173255684142967,
                  },
                  Object {
                    "angle": 0.5711986642890533,
                    "angle0": 0,
                    "color": 2,
                    "label": "apples",
                    "radius": 1,
                    "radius0": 0,
                    "theta": 1,
                    "x": 0.8412535328311812,
                    "y": 0.5406408174555976,
                  },
                ],
              ]
            }
            _colorValue="#12939A"
            _opacityValue={1}
            angleDomain={
              Array [
                0,
                6.283185307179586,
              ]
            }
            arcClassName="rv-radial-chart__series--pie__slice"
            center={
              Object {
                "x": 0,
                "y": 0,
              }
            }
            className="emotion-0"
            clusters={
              Set {
                undefined,
              }
            }
            colorDomain={
              Array [
                0,
                2,
              ]
            }
            colorRange={
              Array [
                "#12939A",
                "#79C7E3",
                "#1A3177",
                "#FF9833",
                "#EF5D28",
              ]
            }
            colorType="category"
            data={
              Array [
                Object {
                  "angle": 6.283185307179586,
                  "angle0": 2.8559933214452666,
                  "color": 0,
                  "label": "cherries",
                  "radius": 1,
                  "radius0": 0,
                  "theta": 6,
                  "x": 1,
                  "y": -2.4492935982947064e-16,
                },
                Object {
                  "angle": 2.8559933214452666,
                  "angle0": 0.5711986642890533,
                  "color": 1,
                  "label": "oranges",
                  "radius": 1,
                  "radius0": 0,
                  "theta": 4,
                  "x": -0.9594929736144974,
                  "y": 0.28173255684142967,
                },
                Object {
                  "angle": 0.5711986642890533,
                  "angle0": 0,
                  "color": 2,
                  "label": "apples",
                  "radius": 1,
                  "radius0": 0,
                  "theta": 1,
                  "x": 0.8412535328311812,
                  "y": 0.5406408174555976,
                },
              ]
            }
            fillDomain={Array []}
            fillRange={
              Array [
                "#EF5D28",
                "#FF9833",
              ]
            }
            getAngle={[Function]}
            getAngle0={[Function]}
            getColor={[Function]}
            getColor0={[Function]}
            getFill={[Function]}
            getFill0={[Function]}
            getLabel={[Function]}
            getOpacity={[Function]}
            getOpacity0={[Function]}
            getRadius={[Function]}
            getRadius0={[Function]}
            getSize={[Function]}
            getSize0={[Function]}
            getStroke={[Function]}
            getStroke0={[Function]}
            getSubLabel={[Function]}
            getX={[Function]}
            getX0={[Function]}
            getY={[Function]}
            getY0={[Function]}
            height={300}
            innerHeight={280}
            innerRadius={120}
            innerWidth={280}
            key=".0"
            labelsRadiusMultiplier={1.1}
            marginBottom={10}
            marginLeft={10}
            marginRight={10}
            marginTop={10}
            onValueClick={[Function]}
            opacity={1}
            opacityDomain={Array []}
            opacityType="literal"
            padAngle={0.02}
            radius={140}
            radiusDomain={
              Array [
                0,
                1,
              ]
            }
            radiusNoFallBack={true}
            radiusRange={
              Array [
                120,
                140,
              ]
            }
            radiusType="linear"
            sameTypeIndex={0}
            sameTypeTotal={1}
            seriesIndex={0}
            sizeDomain={Array []}
            sizeRange={
              Array [
                1,
                10,
              ]
            }
            stack={false}
            strokeDomain={Array []}
            strokeRange={
              Array [
                "#EF5D28",
                "#FF9833",
              ]
            }
            style={Object {}}
            width={300}
            xDomain={
              Array [
                -1,
                1,
              ]
            }
            xRange={
              Array [
                0,
                280,
              ]
            }
            yDomain={
              Array [
                -1,
                1,
              ]
            }
            yRange={
              Array [
                280,
                0,
              ]
            }
          >
            <g
              className="rv-xy-plot__series rv-xy-plot__series--arc emotion-0"
              onClick={[Function]}
              onContextMenu={[Function]}
              onMouseOut={[Function]}
              onMouseOver={[Function]}
              opacity={1}
              pointerEvents="all"
              transform="translate(150,150)"
            >
              <path
                className="rv-xy-plot__series rv-xy-plot__series--arc-path rv-radial-chart__series--pie__slice "
                d="M37.66994874675187,134.83684571146378A140,140,0,1,1,-1.8438781597973435,-139.98785702099957L-1.8438781597973974,-119.98583296927943A120,120,0,1,0,32.03472736865698,115.6450441757701Z"
                key="0"
                onClick={[Function]}
                onContextMenu={[Function]}
                onMouseOut={[Function]}
                onMouseOver={[Function]}
                style={
                  Object {
                    "fill": "#12939A",
                    "opacity": 1,
                    "stroke": "#12939A",
                  }
                }
              />
              <path
                className="rv-xy-plot__series rv-xy-plot__series--arc-path rv-radial-chart__series--pie__slice "
                d="M77.23431846973034,-116.76840347678078A140,140,0,0,1,41.2083250238055,133.7978846945362L35.5731036457106,114.60608315884254A120,120,0,0,0,66.42040783564183,-99.94163007949696Z"
                key="1"
                onClick={[Function]}
                onContextMenu={[Function]}
                onMouseOut={[Function]}
                onMouseOver={[Function]}
                style={
                  Object {
                    "fill": "#1A3177",
                    "opacity": 1,
                    "stroke": "#1A3177",
                  }
                }
              />
              <path
                className="rv-xy-plot__series rv-xy-plot__series--arc-path rv-radial-chart__series--pie__slice "
                d="M1.8438781597973886,-139.98785702099957A140,140,0,0,1,74.13198043765075,-118.76215506798353L63.31806980356222,-101.93538167069971A120,120,0,0,0,1.8438781597973828,-119.98583296927943Z"
                key="2"
                onClick={[Function]}
                onContextMenu={[Function]}
                onMouseOut={[Function]}
                onMouseOver={[Function]}
                style={
                  Object {
                    "fill": "#79C7E3",
                    "opacity": 1,
                    "stroke": "#79C7E3",
                  }
                }
              />
            </g>
          </ArcSeries>
        </svg>
      </div>
    </XYPlot>
  </RadialChart>
</PieChart>
`;
